### PR TITLE
pgadmin4: init at 5.4

### PR DESCRIPTION
**Pull request by mkg20001**
###### Motivation for this change

An attempt to package pgadmin4

NOTE: Currently does not work, but it's pretty close

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cole-h** [comment]
@ofborg eval

(Sorry, ofborg's certs recently had an issue, meaning that it was unable to
react to PRs. This has been fixed, but all PRs filed in this period will need
ofborg to be manually kicked off.)

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -31183,7 +31183,9 @@ in
 
   pgmanage = callPackage ../applications/misc/pgmanage { };
 
-  pgadmin = callPackage ../applications/misc/pgadmin {
+  pgadmin4 = callPackage ../tools/admin/pgadmin { };
+
+  pgadmin3 = callPackage ../applications/misc/pgadmin {
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:31188]: Please move the files to one place and append version numbers to the directories.

File: pkgs/tools/admin/pgadmin/default.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    flask_mail
+    flask_migrate
+    flask_sqlalchemy
+    flask_wtf
+    flask-compress
+    passlib
+    pytz
+    simplejson
+    six
+    (lib.hiPrio speaklater3)
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/pgadmin/default.nix:45]: Why does this need to be hiPrio?

File: pkgs/tools/admin/pgadmin/default.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  };
+
+  # from Dockerfile
+  CPPFLAGS="-DPNG_ARM_NEON_OPT=0";
+
+  format = "other";
+
+  postPatch = ''
+    # sed -r 's|LC_ALL=en_US.UTF-8.* Makefile.sphinx .*$|true|g' -i Makefile
+    patchShebangs .
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/pgadmin/default.nix:25]: Please specify the files/directories that need patching?

File: pkgs/tools/admin/pgadmin/default.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  src = fetchurl {
+    url = "https://ftp.postgresql.org/pub/pgadmin/pgadmin4/v${version}/source/pgadmin4-${version}.tar.gz";
+    sha256 = "N0+fnS0ndH/bqwVMra6kSvKoxRCGPCvLs8aZ0PvWv0o=";
+  };
+
+  # from Dockerfile
+  CPPFLAGS="-DPNG_ARM_NEON_OPT=0";
+
+  format = "other";
+
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/pgadmin/default.nix:22]:
```suggestion
```

File: pkgs/tools/admin/pgadmin/default.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+, lib
+, python3
+, fetchurl
+, zlib
+, python3Packages
+}:
+
+python3Packages.buildPythonApplication rec {
+  pname = "pgadmin";
+  version = "5.4";
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/pgadmin/default.nix:11]:
```suggestion
  version = "5.4";
  format = "other";
```

File: pkgs/development/python-modules/speaklater3/default.nix [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  };
+
+  meta = with lib; {
+    description = "Implements a lazy string for python useful for use with gettext";
+    homepage = "https://github.com/mitsuhiko/speaklater";
+    license = licenses.bsd0;
+    maintainers = with maintainers; [ mkg20001 ];
+  };
+
+}
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/speaklater3/default.nix:22]:
```suggestion
  };
}
```

File: pkgs/development/python-modules/speaklater3/default.nix [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+buildPythonPackage rec {
+  pname = "speaklater3";
+  version = "1.4";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "ySYdS2taMEZyMS0esImb4Cj6yRdgGQqA/szTHTo7UEI=";
+  };
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/speaklater3/default.nix:14]: Missing pythonImportsCheck and checkPhase.

File: pkgs/development/python-modules/httpagentparser/default.nix [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    tox
+  ];
+
+  doCheck = false;
+
+  meta = with lib; {
+    homepage = "https://pypi.org/project/httpagentparser/";
+    description = "Extracts OS Browser etc information from http user agent string";
+    license = licenses.mit;
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/httpagentparser/default.nix:27]:
```suggestion
```

File: pkgs/development/python-modules/httpagentparser/default.nix [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  };
+
+  checkInputs = [
+    tox
+  ];
+
+  doCheck = false;
+
+  meta = with lib; {
+    homepage = "https://pypi.org/project/httpagentparser/";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/httpagentparser/default.nix:24]:
```suggestion
    homepage = "https://shon.github.io/httpagentparser/";
```

File: pkgs/development/python-modules/httpagentparser/default.nix [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "73Y9MZk912GCWs7myLNL4yuVzxZ10cc8PNNfnlKDGyY=";
+  };
+
+  checkInputs = [
+    tox
+  ];
+
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/httpagentparser/default.nix:21]: Why are tests disabled? Can we do a pythonImportsCheck?

File: pkgs/development/python-modules/flask-security-too/default.nix [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+  checkInputs = [
+    pytest
+  ];
+
+  meta = with lib; {
+    homepage = "https://pypi.org/project/Flask-Security-Too/";
+    description = "Simple security for Flask apps (fork)";
+    license = licenses.mit;
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-security-too/default.nix:46]:
```suggestion
```

File: pkgs/development/python-modules/flask-security-too/default.nix [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    email_validator
+    itsdangerous
+    passlib
+    blinker
+  ];
+
+  doCheck = false;
+
+  checkInputs = [
+    pytest
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-security-too/default.nix:39]:
```suggestion
    pytestCheckHook
```

File: pkgs/development/python-modules/flask-security-too/default.nix [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  ];
+
+  meta = with lib; {
+    homepage = "https://pypi.org/project/Flask-Security-Too/";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-security-too/default.nix:43]:
```suggestion
    homepage = "https://github.com/Flask-Middleware/flask-security";
```

File: pkgs/development/python-modules/flask-security-too/default.nix [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    flask_login
+    flask_principal
+    flask_wtf
+    email_validator
+    itsdangerous
+    passlib
+    blinker
+  ];
+
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-security-too/default.nix:36]: Do an pythonImportsCheck?

File: pkgs/development/python-modules/flask-security-too/default.nix [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+, flask_principal
+, passlib
+, email_validator
+, flask_wtf
+, flask_login
+, pytest
+}:
+
+buildPythonPackage rec {
+  pname = "Flask-Security-Too";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-security-too/default.nix:17]:
```suggestion
  pname = "flask-security-too";
```

File: pkgs/development/python-modules/flask-paranoid/default.nix [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    pytest-runner
+  ];
+
+  doCheck = false;
+
+  meta = with lib; {
+    homepage = "https://github.com/miguelgrinberg/flask-paranoid/";
+    description = "Simple user session protection";
+    license = licenses.mit;
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-paranoid/default.nix:34]:
```suggestion
```

File: pkgs/development/python-modules/flask-paranoid/default.nix [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    sha256 = "ou6SV2QuAlTxFXikqlShpyUWemQc+gh5Mw1HfsNwx0k=";
+  };
+
+  propagatedBuildInputs = [
+    flask
+  ];
+
+  checkInputs = [
+    pytest
+    pytest-runner
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-paranoid/default.nix:25]: Please use pytestCheckHook and patchout pytest-runner.

File: pkgs/development/python-modules/flask-paranoid/default.nix [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  ];
+
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-paranoid/default.nix:28]: Why? pythonImportsCheck?

File: pkgs/development/python-modules/flask-paranoid/default.nix [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+, buildPythonPackage
+, fetchPypi
+, pythonAtLeast
+, flask
+, pytest
+, pytest-runner
+}:
+
+buildPythonPackage rec {
+  pname = "Flask-Paranoid";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-paranoid/default.nix:11]:
```suggestion
  pname = "flask-paranoid";
```

File: pkgs/development/python-modules/flask-gravatar/default.nix [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+
+  propagatedBuildInputs = [
+    flask
+  ];
+
+  meta = with lib; {
+    homepage = "https://github.com/zzzsochi/Flask-Gravatar";
+    description = "Small and simple integration of gravatar into flask";
+    license = licenses.bsd3;
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-gravatar/default.nix:35]:
```suggestion
```

File: pkgs/development/python-modules/flask-gravatar/default.nix [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+  };
+
+  # depends on pytest-pep8 which is abandoned, not going to bother
+  doCheck = false;
+
+  preBuild = ''
+    sed "s|tests_require=tests_require,||g" -i setup.py
+    sed "s|extras_require=extras_require,||g" -i setup.py
+    sed "s|setup_requires=setup_requires,||g" -i setup.py
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-gravatar/default.nix:25]: Use postPatch and substituteInPlace.

File: pkgs/development/python-modules/flask-gravatar/default.nix [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+buildPythonPackage rec {
+  pname = "Flask-Gravatar";
+  version = "0.5.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "YGZfMcLGEokdto/4Aek+06CIHGyOw0arxk0qmSP1YuE=";
+  };
+
+  # depends on pytest-pep8 which is abandoned, not going to bother
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-gravatar/default.nix:18]: Are there integration or just liting tests? pythonImportsCheck?

File: pkgs/development/python-modules/flask-gravatar/default.nix [hunk 23]
```diff
@@ -0,0 +1,38 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, pythonAtLeast
+, flask
+, pytest-runner
+}:
+
+buildPythonPackage rec {
+  pname = "Flask-Gravatar";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/flask-gravatar/default.nix:10]:
```suggestion
  pname = "flask-gravatar";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by gador** [comment]
@mkg20001 
I used your PR as a basis and just submitted a PR with a working 6.3 version of pgadmin4. Maybe you find time to have a look at it? Thanks

**Comment by mkg20001** [comment]
Superseded by https://github.com/NixOS/nixpkgs/pull/154764

**Pull request closed without merging**